[Model Averaging] Periodic model averager

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60320 [Model Averaging] Periodic model averager**
* #60303 [Model Averaging] Provide a util function for model averaging

This averager can be used for post-local SGD.

Proposal: https://github.com/pytorch/pytorch/issues/59699
Differential Revision: [D29249850](https://our.internmc.facebook.com/intern/diff/D29249850/)

Commit:
[Model Averaging] Periodic model averager

This averager can be used for post-local SGD.

Differential Revision: [D29249850](https://our.internmc.facebook.com/intern/diff/D29249850/)

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
+            for step in range(0, 20):
+                # Reset the parameters at every step.
+                param.data = copy.deepcopy(tensor)
+                averager.average_parameters(step)
```
I wonder if we can later enhance the API so that the user doesn't have to call `averager.average_parameters` every iteration. For example, can we register a callback on the DDP module that fires every time the parameters need to be averaged?

SciPioneer replied:
It's hard, unless we move optimizer inside DDP. This is because model averaging must occur after `optimizer.step()`, which is actually out of DDP scope. It's a hard boundary at this time.

rohan-varma commented on torch/distributed/algorithms/model_averaging/averagers.py:
```diff
@@ -0,0 +1,91 @@
+            process_group if process_group is not None else dist.group.WORLD
+        )
+
+    def average_parameters(self, step: int):
```
Can we take `step` out of the API and handle accounting for this ourselves? Longer term maybe eliminate this completely and use a hook as I mentioned in another comment

SciPioneer replied:
Are you suggesting add an internal counter which has exactly the same increment as the `step` in the training loop?

I think this is a good suggestion. I have seen some training loops that were implemented across multiple files, so `step` variable may not be very obviously traceable -- to pass `step` to the averager, it may cause some unnecessary user code change.

Let me update the API in a separate PR.

SciPioneer replied:
Addressed in #60320

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
+            "MPI backend does not support creating subgroups on CUDA devices",
+        )
+        @skip_if_lt_x_gpu(2)
+        @skip_if_odd_num_of_gpus()
```
Why does this not work if there is a world size of 5 for example? Can we enhance the test to make it work regardless of world size?

SciPioneer replied:
Yeah, actually I forgot to label this PR [WIP]. I plan to remove `skip_if_odd_num_of_gpus` decorator and simplify the test code.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
+            world_size = dist.get_world_size()
+
+            model = nn.Linear(1, 5, bias=False).cuda(device_id)
+            param = next(model.parameters())
```
any reason we check only one param vs the entire `model.parameters`?

SciPioneer replied:
This is because there is only a single linear layer, so there is only a single parameter tensor.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -976,6 +978,33 @@ def test_average_parameters(self):
+                averager.average_parameters(step)
+                if step < 10 or step % 4 == 0:
+                    # After the model averaging is 0.5
+                    self.assertEqual(param.data, torch.ones_like(param.data) * 0.5)
```
Does it work for larger groups? 

If world_size = 4, then tensor will be 0.5, 1, 1.5, 2 on respective ranks and the average would be 1.25?

SciPioneer replied:
Previously I set the value as `(rank + 1) // 2`, so the value will only be 0 or 1.

Anyway, I updated the test to make it easier to understand.

SciPioneer commented on torch/distributed/algorithms/model_averaging/averagers.py:
```diff
@@ -0,0 +1,91 @@
+        >>>  subgroup, subgroups = dist.new_subgroups()
+        >>>  # Gradients are averaged by each intra-node subgroup during the backward pass.
+        >>>  model = nn.parallel.DistributedDataParallel(
+        >>>     module, device_ids=[rank], output_device=rank, process_group=subgroup
```
Not really, DDP only run allreduce within each node via the subgroup, so there is no global allreduce. In contrast, `averager` runs global allreduce across nodes.

Commit:
Update on "[Model Averaging] Periodic model averager"


This averager can be used for post-local SGD.

Differential Revision: [D29249850](https://our.internmc.facebook.com/intern/diff/D29249850/)

[ghstack-poisoned]

vadimkantorov commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -981,6 +982,32 @@ def test_average_parameters(self):
+
+            model = nn.Linear(1, 5, bias=False).cuda(device_id)
+            param = next(model.parameters())
+            tensor = torch.ones_like(param.data) * rank
```
should this be just `torch.ones_like(param)`?

SciPioneer replied:
`parameter` is not a normal tensor, but has  both `.data` and `.grad` fields.

vadimkantorov commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -981,6 +982,32 @@ def test_average_parameters(self):
+            model = nn.Linear(1, 5, bias=False).cuda(device_id)
+            param = next(model.parameters())
+            tensor = torch.ones_like(param.data) * rank
+            expected_avg_tensor = torch.ones_like(param.data) * sum(range(world_size)) / world_size
```
same here: `torch.ones_like(param)`?

rohan-varma reviewed (approved):
Awesome!
